src/gtk/toga_gtk/widgets/detailedlist.py

```diff
@@ -1,92 +1,185 @@
-        self.native.add(self.treeview)
-        self.native.set_min_content_width(self.interface.MIN_WIDTH)
-        self.native.set_min_content_height(self.interface.MIN_HEIGHT)
+        self.store = Gio.ListStore()
+        self.list_box.bind_model(self.store, lambda a: a)
 
-    def change_source(self, source):
-        self.treeview.set_model(None)
-        self.store.change_source(source)
-        for i, node in enumerate(source):
-            self.insert(i, node)
+        self.scrolled_window = Gtk.ScrolledWindow()
 
-        self.treeview.set_model(self.store)
+        self.scrolled_window.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.AUTOMATIC)
+        self.scrolled_window.set_min_content_width(self.interface.MIN_WIDTH)
+        self.scrolled_window.set_min_content_height(self.interface.MIN_HEIGHT)
 
-    def insert(self, index, item):
-        self.store.insert(item)
+        self.scrolled_window.add(self.list_box)
 
-    def change(self, item):
-        self.store.change(item)
+        self.refresh_button = RefreshButton(self.scrolled_window.get_vadjustment())
 
-    def remove(self, item, index):
-        self.store.remove(item, index)
+        self.scroll_button = ScrollButton(self.scrolled_window.get_vadjustment())
+        self.scroll_button.set_scroll(lambda: self.scroll_to_row(-1))
+        
+        self.native = Gtk.Overlay()
+        self.native.add_overlay(self.scrolled_window)
 
-    def clear(self):
-        self.store.clear()
+        self.refresh_button.overlay_over(self.native)
+        self.scroll_button.overlay_over(self.native)
 
-    def set_on_refresh(self, handler):
-        pass
+        self.native.interface = self.interface
 
-    def after_on_refresh(self):
-        # No special handling required
-        pass
+        self._on_select_signal_handler = self.list_box.connect(
+            'row-selected', lambda w, item: self._on_select(item))
+
+        self.right_click_gesture = Gtk.GestureMultiPress.new(self.list_box)
+        self.right_click_gesture.set_button(3)
+        self.right_click_gesture.set_propagation_phase(Gtk.PropagationPhase.BUBBLE)
+        self.right_click_gesture.connect("pressed", self._on_right_click)
+        
+    @property
+    def on_delete(self):
+        if self._on_delete_handler is not None:
+            return self._on_delete
+        return None
+
+    def row_factory(self, item: 'Row'):
+        new_item = TextIconRow(self.interface.factory, self, item)
+        setattr(item, '_impl', new_item)
+        return new_item
+
+    def destroy(self):
+        self.disconnect(self._on_select_signal_handler)
+        super().destroy()
+
+    def change_source(self, source: 'ListSource'):
+        self.store.remove_all()
+        for item in source:
+            self.store.append(
+                self.row_factory(item))
+
+        # We have to wait until the rows are actually added decide how to position the
+        # refresh button
+        GLib.idle_add(lambda: not self._changed())
```

Comment:
What does this do? It looks like it's adding a global run loop callback... which I would have thought would come with a performance hit.

Reply:
Display of the scroll and refresh buttons depends on how many rows there are on the list, or more specifically on the dimensions of the list. 

We can't know the dimensions of each row (and thus of the list) until gtk allocates space for it. Gtk does emit `size-allocate` after allocation but I couldn't find any guarantees that the rows have their sizes allocated in the order they are inserted in the `ListStore` and in my opinion that's unlikely to be the case. 

Therefore we would need to wait for `size-allocate` on all rows and either update the visibility of the buttons on all `size-allocates` or have a counter and only do it on the last `size-allocate`. Obviously none of those options are desirable.

Fortunately functions added with `idle_add` are run when gtk is idle and thus after any size allocation. This solves our problem and from the perspective of the user this happens immediately.

With respect to the performance hit: even though we are adding the callback to the global loop, it only runs once. This is what that lambda is for. If a callback returns `False` then it's not ran again. I used a lambda because returning `False` from `self._changed()` would mean returning `False` on success.

I'm pretty sure I had a comment explaining this. Guess I deleted it by accident? Should I add the text above as a comment?

Reply:
That would definitely be helpful. The combination of the lambda, the GTK handling of a `false` return value, and the slightly opaque name `_changed` means it's not at all clear that this will be invoked once.